llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,432 @@
+  unsigned int ArgNo = 0;
+  for (Function::arg_iterator I = F.arg_begin(), E = F.arg_end(); I != E;
+       ++I, ++I2, ArgNo++) {
+    auto ArgIt = ToModify.find(I);
+    if (ArgIt == ToModify.end()) {
+      // Transfer old arguments as is
+      I->replaceAllUsesWith(I2);
+      I2->takeName(I);
+    }
+  }
+
+  // Fix call sites
+  SmallVector<std::pair<Instruction *, Instruction *>, 10> OldNewInst;
+  for (auto &use : F.uses()) {
+    // Use must be a call site
+    SmallVector<Value *, 10> Params;
+    auto User = use.getUser();
+    if (isa<CallInst>(User)) {
+      auto Call = cast<CallInst>(User);
+      for (unsigned int I = 0,
+                        NumOpnds = cast<CallInst>(Call)->getNumArgOperands();
+           I != NumOpnds; I++) {
+        auto SrcOpnd = Call->getOperand(I);
+        auto NewTy = argIsSimdPtr(SrcOpnd);
+        if (NewTy) {
+          auto BitCast = new BitCastInst(SrcOpnd, NewTy, "", Call);
+          Params.push_back(BitCast);
+        } else {
+          if (SrcOpnd != &F)
+            Params.push_back(SrcOpnd);
+          else
+            Params.push_back(NF);
+        }
+      }
+
+      // create new call instruction
+      auto NewCallInst = CallInst::Create(NFTy, NF, Params, "");
+      NewCallInst->setCallingConv(F.getCallingConv());
+      OldNewInst.push_back(std::make_pair(Call, NewCallInst));
+    }
+  }
+
+  for (auto InstPair : OldNewInst) {
+    auto OldInst = InstPair.first;
+    auto NewInst = InstPair.second;
+    ReplaceInstWithInst(OldInst, NewInst);
+  }
+
+  F.eraseFromParent();
+
+  return NF;
+}
+
+bool ESIMDLowerVecArgPass::hasGlobalConstExpr(Value *V,
+                                              GlobalVariable *&Global) {
+  if (isa<GlobalVariable>(V)) {
+    Global = cast<GlobalVariable>(V);
+    return true;
+  }
+
+  if (isa<ConstantExpr>(V)) {
+    auto FirstOpnd = cast<ConstantExpr>(V)->getOperand(0);
+    return hasGlobalConstExpr(FirstOpnd, Global);
+  }
+
+  return false;
+}
+
+// Replace ConstantExpr if it contains old global variable.
+ConstantExpr *
+ESIMDLowerVecArgPass::createNewConstantExpr(GlobalVariable *NewGlobalVar,
+                                            Type *OldGlobalType, Value *Old) {
+  ConstantExpr *NewConstantExpr = nullptr;
+
+  if (isa<GlobalVariable>(Old)) {
+    NewConstantExpr = cast<ConstantExpr>(
+        ConstantExpr::getBitCast(NewGlobalVar, OldGlobalType));
+    return NewConstantExpr;
+  }
+
+  auto InnerMost = createNewConstantExpr(
+      NewGlobalVar, OldGlobalType, cast<ConstantExpr>(Old)->getOperand(0));
+
+  NewConstantExpr = cast<ConstantExpr>(
+      cast<ConstantExpr>(Old)->getWithOperandReplaced(0, InnerMost));
+
+  return NewConstantExpr;
+}
+
+// Globals are part of ConstantExpr. This loop iterates over
+// all such instances and replaces them with a new ConstantExpr
+// consisting of new global vector* variable.
+void ESIMDLowerVecArgPass::replaceConstExprWithGlobals(Module &M) {
+  for (auto &F : M) {
```

Comment:
This seems like a very slow way of finding global variables in const expressions. How about using `for (auto &GV : M.globals())` and then iterate over uses of the global variables?

Reply:
fixed.